test/parallel/test-crypto-keygen.js

```diff
@@ -0,0 +1,631 @@
+'use strict';
+
+const common = require('../common');
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const assert = require('assert');
+const {
+  createSign,
+  createVerify,
+  generateKeyPair,
+  generateKeyPairSync,
+  publicEncrypt,
+  privateDecrypt
+} = require('crypto');
+
+// Asserts that the size of the given key (in chars or bytes) is within 10% of
+// the expected size.
+function assertApproximateSize(key, expectedSize) {
+  const u = typeof key === 'string' ? 'chars' : 'bytes';
+  const min = Math.floor(0.9 * expectedSize);
+  const max = Math.ceil(1.1 * expectedSize);
+  assert(key.length >= min,
+         `Key (${key.length} ${u}) is shorter than expected (${min} ${u})`);
+  assert(key.length <= max,
+         `Key (${key.length} ${u}) is longer than expected (${max} ${u})`);
+}
+
+// Tests that a key pair can be used for encryption / decryption.
+function testEncryptDecrypt(publicKey, privateKey) {
+  const message = 'Hello Node.js world!';
+  const plaintext = Buffer.from(message, 'utf8');
+  const ciphertext = publicEncrypt(publicKey, plaintext);
+  const received = privateDecrypt(privateKey, ciphertext);
+  assert.strictEqual(received.toString('utf8'), message);
+}
+
+// Tests that a key pair can be used for signing / verification.
+function testSignVerify(publicKey, privateKey) {
+  const message = 'Hello Node.js world!';
+  const signature = createSign('SHA256').update(message)
+                                        .sign(privateKey, 'hex');
+  const okay = createVerify('SHA256').update(message)
+                                     .verify(publicKey, signature, 'hex');
+  assert(okay);
+}
+
+// Constructs a regular expression for a PEM-encoded key with the given label.
+function getRegExpForPEM(label) {
+  const head = `\\-\\-\\-\\-\\-BEGIN ${label}\\-\\-\\-\\-\\-`;
+  const body = '([a-zA-Z0-9\\+/=]{64}\n)*[a-zA-Z0-9\\+/=]{1,64}';
+  const end = `\\-\\-\\-\\-\\-END ${label}\\-\\-\\-\\-\\-`;
+  return new RegExp(`^${head}\n${body}\n${end}\n$`);
+}
+
+const pkcs1PubExp = getRegExpForPEM('RSA PUBLIC KEY');
+const pkcs1PrivExp = getRegExpForPEM('RSA PRIVATE KEY');
+const spkiExp = getRegExpForPEM('PUBLIC KEY');
+const pkcs8Exp = getRegExpForPEM('PRIVATE KEY');
+const pkcs8EncExp = getRegExpForPEM('ENCRYPTED PRIVATE KEY');
+const sec1Exp = getRegExpForPEM('EC PRIVATE KEY');
+
+// Since our own APIs only accept PEM, not DER, we need to convert DER to PEM
+// for testing.
+function convertDERToPEM(label, der) {
+  const base64 = der.toString('base64');
+  const lines = [];
+  let i = 0;
+  while (i < base64.length) {
+    const n = Math.min(base64.length - i, 64);
+    lines.push(base64.substr(i, n));
+    i += n;
+  }
+  const body = lines.join('\n');
+  const r = `-----BEGIN ${label}-----\n${body}\n-----END ${label}-----\n`;
+  assert(getRegExpForPEM(label).test(r));
+  return r;
+}
```

Comment:
maybe this should be part of the API as well? I think Node should provide a way to read this format, if we generate it.